### PR TITLE
Add GitHubEntAuth object

### PR DESCRIPTION
**Pull request by dragon512**
## Remove this paragraph
Please have a look at our developer documentation before submitting your Pull Request.

http://trac.buildbot.net/wiki/Development
And especially:
http://trac.buildbot.net/wiki/SubmittingPatches

## Contributor Checklist:

* [x] I have updated the unit tests
* [x] I have created a file in the master/buildbot/newsfragment directory (and read the README.txt in that directory)
* [x] I have updated the appropriate documentation

### CODE REVIEW

**Comment by anish** [comment]
@dragon512 can you include sample documentation on how to use this ?

File: master/buildbot/www/oauth2.py [hunk 1]
```diff
@@ -188,6 +188,35 @@ def getUserInfoFromOAuthClient(self, c):
                     email=data["email"],
                     avatar_url=data["picture"])
 
+class GitHubEntAuth(OAuth2Auth):
+    def __init__(self,
```
Comment (anish) [inline master/buildbot/www/oauth2.py]: can you not extend GitHubAuth instead ? the github reporter uses the same code for both : https://github.com/buildbot/buildbot/blob/58ac4c7adc1896842c70f9dee1c85a479da95a08/master/buildbot/reporters/github.py#L47

Comment (dragon512) [inline master/buildbot/www/oauth2.py]: I should have just added some updated documentation with an example

File: master/buildbot/www/oauth2.py [hunk 2]
```diff
@@ -197,14 +196,29 @@ class GitHubAuth(OAuth2Auth):
...
+                 clientId, clientSecret, serverURL=None, autologin=False, **kwargs):
+
+        OAuth2Auth.__init__(self, clientId, clientSecret, autologin, **kwargs)
+        if serverURL is not None:
+            # setup for enterprise github
+            if not serverURL.endswith("/"):
+                serverURL = serverURL + "/"
+
+            GitHubAuth.authUri = '{0}login/oauth/authorize'.format(serverURL)
```
Comment (tardyp) [inline master/buildbot/www/oauth2.py]: self.authUri

File: master/buildbot/www/oauth2.py [hunk 3]
```diff
@@ -197,14 +196,29 @@ class GitHubAuth(OAuth2Auth):
...
+            GitHubAuth.tokenUri = '{0}login/oauth/access_token'.format(serverURL)
+            GitHubAuth.resourceEndpoint = '{0}api/v3/'.format(serverURL)
+
     def getUserInfoFromOAuthClient(self, c):
-        user = self.get(c, '/user')
-        emails = self.get(c, '/user/emails')
+        map_data = c.get(self.resourceEndpoint).json()
```
Comment (tardyp) [inline master/buildbot/www/oauth2.py]: I dont think you really need to fetch first this mapping.
This root endpoint is for discoverability, but the resource endpoints are actually stable, so I think you can keep the old implementation, and only keep the change in __init__

Comment (dragon512) [inline master/buildbot/www/oauth2.py]: Let me double check this.. I believe had a dump and the URL was from for enterprise git, So I had changed it to use the mapping as it was intended. I will try to do this tomorrow and get an update

File: master/buildbot/www/oauth2.py [hunk 4]
```diff
@@ -197,6 +196,19 @@ class GitHubAuth(OAuth2Auth):
...
+
+        OAuth2Auth.__init__(self, clientId, clientSecret, autologin, **kwargs)
+        if serverURL is not None:
+            # setup for enterprise github
+            if serverURL.endswith("/"):
+                serverURL = serverURL[:-1]
+
+            GitHubAuth.authUri = '{0}/login/oauth/authorize'.format(serverURL)
+            GitHubAuth.tokenUri = '{0}/login/oauth/access_token'.format(serverURL)
+            GitHubAuth.resourceEndpoint = '{0}/api/v3'.format(serverURL)
```
Comment (tardyp) [inline master/buildbot/www/oauth2.py]: Still need to change to self.xxx (dont change the class variables, change the instance variables)
then please add a unit test to keep the test coverage high.
https://docs.codecov.io/docs/browser-extension

**Comment by dragon512** [comment]
will you want me to squash these?

**Comment by tardyp** [comment]
yes, please squash

**Comment by dragon512** [comment]
Anything else? I think I have all the bases covered.

File: master/docs/manual/cfg-www.rst [hunk 5]
```diff
@@ -238,6 +239,14 @@ The available classes are described here:
             'auth': util.GitHubAuth("clientid", "clientsecret"),
         }
 
+    Example for Enterprise GitHub::
+
+        from buildbot.plugins import util
+        c['www'] = {
+            # ...
+            'auth': util.GitHubAuth("clientid", "clientsecret","https://git.corp.mycompany.com"),
```
Comment (seankelly) [inline master/docs/manual/cfg-www.rst]: Please add a space after `"clientsecret",`.

File: master/buildbot/newsfragments/githubauth.feature [hunk 6]
```diff
@@ -0,0 +1 @@
+Changed :py:class:`~buildbot.www.oauth2.GitHubAuth` now supports GitHub Enterprise when setting new serverURL argument.
```
Comment (tardyp) [inline master/buildbot/newsfragments/githubauth.feature]: please quote serverURL

    ``serverURL``

**Comment by dragon512** [comment]
I think everything here should be good. Let me know if you need other tweak

**Comment by tardyp** [comment]
thanks!

**Pull request merged**